### DISCUSSION TITLE
Options to pause and stop sound files in dialog.tlk

### DISCUSSION THREAD

**Discussion by Incrementis** [NearInfinityBrowser/NearInfinity#189] [Ideas & Feature Requests]
When setting a strref with an associated sound file, I listened to a slightly longer sound file, namely “_CHAPTER1.WAV_”. The sound file lasts 24 seconds.
There is no option to pause or stop the sound, as is the case with directly selecting a sound file in the WAV folder.

Since I'm working on a mod that has sound files that can reach for a strref minutes, being able to pause and stop the sound would be beneficial in this case.

**Example for :**
- _pause:_ Something more important is happening and I have to pause for an indefinite amount of time
- _stop:_ I just want to listen to this strref sound within dialog.tlk for a short time to check if the file contains the correct content

**Suggestion:** See screenshot.
![grafik](https://github.com/user-attachments/assets/9f71934d-1e32-4b27-83a2-fe75fb6e0d05)

EDIT:
Version: v2.4-20240914

**Comment by Argent77** [comment]
I agree, that would be a useful improvement. I'll add it to my todo list.

In the meantime you could open the WAV resource view with the `View/Edit` button and use its playback controls.

**Comment by Argent77** [comment]
@Incrementis I have taken the opportunity to implement a new class for playing sound clips. It is currently enabled for the WAV resource view and the resource reference datatype that lists WAV resources (including the sound reference in the string table).

Could you please test this feature and see if you can find ways to "break" it or make it behave in unexpected ways? It's an implementation from scratch which may still have some bugs left. (A known issue is that sound playback in the String Editor won't stop automatically when the window is closed.)

I would also be interested if all the associated UI elements (time labels, progress slider, loop checkbox, etc.) are suitable for the situation where it is displayed. The player is customizable, so I can easily add or remove UI elements from view.

Download: https://github.com/Argent77/NearInfinity/releases/tag/nightly

**Comment by Incrementis** [comment]
Yes, I can do that. I will report here.

**Comment by Incrementis** [comment]
## Here are some results(BG2:EE):

### First rudimentary impressions (possibly just a small challenge)
- The existence of the pause button is not obvious
- Found the pause button by pressing the play button after the play button was activated

This could be prevented by changing the shape of the Play button to a Pause button after pressing the Play button.
This concept can also be found in YouTube videos.

The time bar has more of a psychological effect in my opinion as seeing the movement and progress gives a sense of control. If you want to deepen the psychological effect of control, you can incorporate the waveform of the sound files (see screenshot as an example), which should also have a moving timeline that moves parallel to the manipulative time bar.
![grafik](https://github.com/user-attachments/assets/c5f818b8-1361-404a-b09d-c05f2a39b0ed)
The disadvantage is that the implementation can become complicated when the length of the audio file reaches minutes instead of seconds (visual representation of the waveform). The decision is yours. Personally I don't see any use case other than the psychological effect (but I could be wrong here).

### Test1
- Sounfile(WAV folder): 2YAGA23 [11,8 seconds]

1. Activate loop checkbox
2. Press Play buton
3. Click on the timer bar repeatedly and relatively quickly (see image for reference).
![grafik](https://github.com/user-attachments/assets/66670efc-70c9-4490-b201-9e680e5a8e2e)

The loop doesn't work even though it is enabled. If step 3. is not done, the loop works as expected.

### Test2
- Sounfile(WAV folder): 2YAGA23 [11,8 seconds]

1. Loop checkbox is unchecked
2. Press Play button
3. Press Play button to pause
4. Activate loop checkbox

After step 4, the sound will be played automatically. The play button still shows as pause and the timer bar is frozen.
This issue also occurs if the Loop checkbox is checked at startup and then unchecked in step 4. In this case see screenshot for status after sound file reaches end. The stop button is also still available. Pressing it will put everything back in order.
![grafik](https://github.com/user-attachments/assets/1a4601db-fe0f-41ef-bc91-9b5c6b38d636)
In this state, if the loop checkbox is checked, the sound will start playing again.

### Test3
- Sounfile(WAV folder): 2YAGA23 [11,8 seconds]

1. Loop checkbox is unchecked
2. Press Play button
3. Press Play button to pause
4. Activate loop checkbox (this will trigger issue found in Test2)
5. Press Stop button 

This will restart the sound file and create an infinite loop since unchecking the loop checkbox does not stop playback.
Pressing the Play button twice helps here, as this activates the availability of the Stop button.

**Reply by Argent77** [reply]
Thank you. That's very helpful.

Setting up loop mode internally is very peculiar. It's easy to miss or misconfigure code paths.

Combining play and pause in a single button was mainly done for space reasons. I'll see if there is still room for improvements.

**Comment by Argent77** [comment]
Reported issues should be fixed in the latest nightly.

Playback controls have been improved. WAV resource view provides separate play and pause buttons. Sound playback panel on resource lists uses combined play/pause button for space reasons, but I added a tooltip with further info.

**Comment by Incrementis** [comment]
There are some visual inconsistencies regarding the progress slider. The inconsistencies may be seen as outliers in some cases, but I will use forum thread examples to show that there is indeed interest in longer recordings (besides my personal interest), which will be addressed at the end of this comment.
I don't know if the problem will be solvable, but it will at least be enlightening to see.

**OS:**
- Windows 11

**Scaling:** 
- 125%

**Screen resolution during testing:**
- 2560 x1440
- NI is used in full screen mode

**Game:**
- BG:EE(SoD)

# Visual Examples(from OGG files from self-made recordings)
## Comparing over 2 minutes length

![!_HOFS08_2min_18sec](https://github.com/user-attachments/assets/b80b680b-13e1-4c34-935d-94aa0dd93e6c)
**strref:** 13154
[Download file !_HOFS08.OGG](https://github.com/Incrementis/Tutorils-Historical-Encyclopedias-/raw/refs/heads/main/tutorils_he/sound/english/shadowdale/!_HOFS08.OGG)


![!_HOFS00_2min_47sec](https://github.com/user-attachments/assets/b124f6ec-8839-4864-b55b-231ff0e56c9f)
**strref:** 13146
[Download file !_HOFS00.OGG](https://github.com/Incrementis/Tutorils-Historical-Encyclopedias-/raw/refs/heads/main/tutorils_he/sound/english/shadowdale/!_HOFS00.OGG)

**Note:** The time symbols in the display overlap.

## Comparing over 3 minutes length

![!_HOFW02_3min_33sec](https://github.com/user-attachments/assets/2c72b628-c639-4cf6-a868-e35ce0b4e1f4)
**strref:** 13190
[Download file !_HOFW02.OGG](https://github.com/Incrementis/Tutorils-Historical-Encyclopedias-/raw/refs/heads/main/tutorils_he/sound/english/waterdeep/!_HOFW02.OGG)

![!_HOFW04_3min_45sec](https://github.com/user-attachments/assets/e39dea1a-753c-495a-bfba-29ea8a86fba0)
**strref:** 13192
[Download file !_HOFW04.OGG](https://github.com/Incrementis/Tutorils-Historical-Encyclopedias-/raw/refs/heads/main/tutorils_he/sound/english/waterdeep/!_HOFW04.OGG)

![!_HOTN04_3min_51sec](https://github.com/user-attachments/assets/d6f9f8cd-45dd-4351-9d5a-4a532447a6f4)
This one looks almost as expected.
**strref:** 13174
[Download file !_HOTN04.OGG](https://github.com/Incrementis/Tutorils-Historical-Encyclopedias-/raw/refs/heads/main/tutorils_he/sound/english/north/!_HOTN04.OGG)

**Note:** This time the length is more than 2 minutes, although there are not overlaps in all instances.

## Length of over 12 min(Rare)

![!_HOTN07_12min_40sec](https://github.com/user-attachments/assets/af071c7e-82fb-4415-85a1-245d6f0c843a)
**strref:** 13177
[Download file !_HOTN07.OGG](https://github.com/Incrementis/Tutorils-Historical-Encyclopedias-/raw/refs/heads/main/tutorils_he/sound/english/north/!_HOTN07.OGG)

**Note:** This recording is actually from the book "BOOK50.ITM - History of the North VIII" and has as far as I can tell compared to the other books a large character count(image below: notepad++)
![grafik](https://github.com/user-attachments/assets/d49ceb7e-e826-45ee-996f-b959c57c092a)

I also think the overlap might be hard to resolve at some point regarding the time length of a sound file.

# Interest in longer recordings:

- [[BG1] Voices Voices Extravaganza - AI dialog voiceovers - The Gibberlings3](https://www.gibberlings3.net/forums/topic/38809-bg1-voices-voices-extravaganza-ai-dialog-voiceovers/#comment-343683)
- [On Reddit: Working on an accessibility mod for the visually impaired and would like your thoughts on how to proceed with voiced text of the important characters!](https://www.reddit.com/r/baldursgate/comments/1gbg42m/working_on_an_accessibility_mod_for_the_visually/?rdt=40600)

**Comment by Argent77** [comment]
Some overlapping is situational and cannot really be avoided (without creating other side effects). Time labels are applied in regular intervals as well as at the beginning and end of a sound clip. An overlap may occur if a regular interval and the end of the sound clip are in close proximity.

Time label intervals already depend somewhat on the sound clip length, but I only took vanilla clips into account for interval calculations. I'll see if I can fine-tune it further for clip lengths of 10 minutes or more. But it cannot be extended indefinitely since the interval thresholds are implemented manually.
